### PR TITLE
Feat/separate servers

### PR DESCRIPTION
**Pull request by gabisavu-ionos**
## Checklist

- [x] PR name added as appropriate (e.g. `feat:`/`fix:`/`doc:`/`test:`/`refactor:`)
- [ ] Tests added or updated
- [x] Documentation updated
- [ ] Changelog updated and version incremented (label: upcoming release)
- [x] Github Issue linked if any
- [x] Jira task updated

### CODE REVIEW

File: ionoscloud/resource_server.go [hunk 1]
```diff
@@ -392,12 +392,16 @@ func checkServerImmutableFields(_ context.Context, diff *schema.ResourceDiff, _
 func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client := meta.(SdkBundle).CloudApiClient
 
-	server, volume, err := initializeCreateRequests(d)
-
-	if err != nil {
-		diags := diag.FromErr(err)
-		return diags
+	//server, volume, err := initializeCreateRequests(d)
+	server := ionoscloud.Server{
```
Comment (iblindu) [inline ionoscloud/resource_server.go:396]: Also in create, read and update you have to remove the template_uuid field

File: ionoscloud/resource_server.go [hunk 2]
```diff
@@ -1250,31 +1254,7 @@ func resourceServerDelete(ctx context.Context, d *schema.ResourceData, meta inte
...
-		// Wait, catching any errors
-		_, errState := getStateChangeConf(meta, d, apiResponse.Header.Get("Location"), schema.TimeoutDelete).WaitForStateContext(ctx)
-		if errState != nil {
-			diags := diag.FromErr(fmt.Errorf("error getting state change for volumes delete %w", errState))
-			return diags
-		}
-	}
-
-	apiResponse, err = client.ServersApi.DatacentersServersDelete(ctx, dcId, d.Id()).Execute()
+	apiResponse, err := client.ServersApi.DatacentersServersDelete(ctx, dcId, d.Id()).Execute()
```
Comment (iblindu) [inline ionoscloud/resource_server.go:1256]: why did you remove all this? are you sure that the inline volume is also deleted?

File: ionoscloud/resource_server_cube.go [hunk 3]
```diff
@@ -0,0 +1,751 @@
...
+		ReadContext:   resourceServerRead,
+		UpdateContext: resourceServerUpdate,
+		DeleteContext: resourceCubeServerDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceServerImport,
+		},
+		CustomizeDiff: checkServerImmutableFields,
+
+		Schema: map[string]*schema.Schema{
+			// Cybe Server parameters
```
Comment (iblindu) [inline ionoscloud/resource_server_cube.go]: cube server*

File: ionoscloud/resource_server_cube.go [hunk 4]
```diff
@@ -0,0 +1,751 @@
...
+
+	var sshKeyPath []interface{}
+	var publicKeys []string
+	var image, imageAlias, imageInput string
+	var isSnapshot bool
+	var diags diag.Diagnostics
+	var password, licenceType string
+
+	dcId := d.Get("datacenter_id").(string)
+
```
Comment (iblindu) [inline ionoscloud/resource_server_cube.go:355]: where are you taking the template_uuid?

**Review by iblindu — CHANGES_REQUESTED** [review]

**Comment by iblindu** [comment]
You don't have a data-source, import and test for cube servers

**Review by iblindu — CHANGES_REQUESTED** [review]
You don't have a data-source, import and test file for cube servers.

File: ionoscloud/resource_server.go [hunk 5]
```diff
@@ -32,7 +32,7 @@ func resourceServer() *schema.Resource {
 			// Server parameters
 			"template_uuid": {
 				Type:     schema.TypeString,
-				Optional: true,
+				Required: true,
```
[the remark this answers is not in the document]
Comment (gabisavu-ionos) [reply ionoscloud/resource_server.go:35]: no, is for both

File: ionoscloud/resource_server.go [hunk 6]
```diff
@@ -392,12 +392,16 @@ func checkServerImmutableFields(_ context.Context, diff *schema.ResourceDiff, _
 func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client := meta.(SdkBundle).CloudApiClient
 
-	server, volume, err := initializeCreateRequests(d)
-
-	if err != nil {
-		diags := diag.FromErr(err)
-		return diags
+	//server, volume, err := initializeCreateRequests(d)
```
[the remark this answers is not in the document]
Comment (gabisavu-ionos) [reply ionoscloud/resource_server.go:395]: added

File: ionoscloud/resource_server_cube.go (hunk 3 above)
Comment (gabisavu-ionos) [reply ionoscloud/resource_server_cube.go]: changed

File: ionoscloud/resource_server_cube.go (hunk 4 above)
Comment (gabisavu-ionos) [reply ionoscloud/resource_server_cube.go:355]: oh, I have an unpushed commit, sry

File: ionoscloud/resource_server.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gabisavu-ionos) [reply ionoscloud/resource_server.go]: yes, but I don't know if we will decide to remove it directly, or keep it for a while and mark as deprecated

File: ionoscloud/constants.go [hunk 7]
```diff
@@ -136,6 +136,7 @@ const (
 // Server Constants
 const (
 	ServerResource         = "ionoscloud_server"
+	ServerCubeResource     = "ionoscloud_cube_server"
```
[the remark this answers is not in the document]
Comment (gabisavu-ionos) [reply ionoscloud/constants.go:139]: will be added lately

File: ionoscloud/provider.go [hunk 8]
```diff
@@ -69,6 +69,7 @@ func Provider() *schema.Provider {
 			"ionoscloud_loadbalancer":   resourceLoadbalancer(),
 			NicResource:                 resourceNic(),
 			ServerResource:              resourceServer(),
+			ServerCubeResource:          resourceCubeServer(),
```
[the remark this answers is not in the document]
Comment (gabisavu-ionos) [reply ionoscloud/provider.go:72]: the adta sources will be split also

File: ionoscloud/data_source_cube_server.go [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+			},
+			"id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			//"type": {
```
Comment (cristiGuranIonos) [inline ionoscloud/data_source_cube_server.go:34]: remove comment

File: ionoscloud/resource_server_cube.go [hunk 10]
```diff
@@ -0,0 +1,1075 @@
...
+		if err := d.Set("firewallrule_id", parts[3]); err != nil {
+			return nil, fmt.Errorf("error setting firewallrule_id %w", err)
+		}
+	}
+	d.SetId(parts[1])
+
+	return []*schema.ResourceData{d}, nil
+}
+
+// nu are cores, ram si volume.size cube sererul
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_server_cube.go:1075]: nu are :)

File: ionoscloud/data_source_cube_server.go [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     nicServerDSResource,
+			},
+		},
+		Timeouts: &resourceDefaultTimeouts,
+	}
+}
+
+func setCubeServerData(d *schema.ResourceData, server *ionoscloud.Server, token *ionoscloud.Token) error {
```
Comment (cristiGuranIonos) [inline ionoscloud/data_source_cube_server.go:174]: maybe make a setServerData that has the common stuff, then a setEnterpriseData and setCubeData that have only the cube and enterprise parts. That way we don't get duplicated code.

File: ionoscloud/resource_server_cube.go [hunk 12]
```diff
@@ -0,0 +1,1075 @@
...
+			}
+		}
+
+		if server.Properties.Name != nil {
+			if err := d.Set("template_uuid", *server.Properties.TemplateUuid); err != nil {
+				return nil, fmt.Errorf("error setting template uuid %w", err)
+			}
+		}
+
+		//if server.Properties.Cores != nil {
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_server_cube.go:955]: comment

File: ionoscloud/resource_server_cube.go [hunk 13]
```diff
@@ -0,0 +1,1075 @@
...
+	if err != nil {
+		diags := diag.FromErr(fmt.Errorf("an error occured while deleting a server ID %s %w", d.Id(), err))
+		return diags
+
+	}
+
+	// Wait, catching any errors
+	_, errState := getStateChangeConf(meta, d, apiResponse.Header.Get("Location"), schema.TimeoutDelete).WaitForStateContext(ctx)
+	if errState != nil {
+		diags := diag.FromErr(fmt.Errorf("error getting state change for datacenter delete %w", errState))
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_server_cube.go:896]: for cube server delete

File: ionoscloud/resource_server_cube.go [hunk 14]
```diff
@@ -0,0 +1,1075 @@
...
+				},
+			},
+		},
+	}
+	var primaryNic *ionoscloud.Nic
+	if server.Entities.Nics != nil && server.Entities.Nics.Items != nil && len(*server.Entities.Nics.Items) > 0 {
+		primaryNic = &(*server.Entities.Nics.Items)[0]
+	}
+	// Nic Arguments
+	if _, ok := d.GetOk("nic"); ok {
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_server_cube.go:507]: should be the same for server I think

File: ionoscloud/resource_server_cube.go [hunk 15]
```diff
@@ -0,0 +1,1336 @@
...
+		Schema: map[string]*schema.Schema{
+			// Cube Server parameters
+			"template_uuid": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.All(validation.StringIsNotWhiteSpace),
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_server_cube.go:37]: use validateDiagFunc instead

File: docs/data-sources/cube_server.md [hunk 16]
```diff
@@ -0,0 +1,124 @@
+---
+subcategory: "Compute Engine"
+layout: "ionoscloud"
+page_title: "IonosCloud : ionoscloud_server"
+sidebar_current: "docs-ionoscloud-datasource-server"
+description: |-
+  Get information on a Ionos Cloud Servers
```
Comment (cristiGuranIonos) [inline docs/data-sources/cube_server.md:7]: cube

File: docs/data-sources/cube_server.md [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+---
+
+# ionoscloud\_cube\_server
+
+The **Server data source** can be used to search for and return existing servers.
```
Comment (cristiGuranIonos) [inline docs/data-sources/cube_server.md:12]: Cube server data source

File: docs/resources/cube_server.md [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+	location        = "de/txl"
+}
+
+resource "ionoscloud_lan" "example" {
+  datacenter_id     = ionoscloud_datacenter.example.id
+  public            = true
+  name              = "Lan Example"
+}
+
+resource "ionoscloud_server" "example" {
```
Comment (cristiGuranIonos) [inline docs/resources/cube_server.md:36]: this is a server example

File: docs/resources/cube_server.md [hunk 19]
```diff
@@ -0,0 +1,98 @@
+---
+subcategory: "Compute Engine"
+layout: "ionoscloud"
+page_title: "IonosCloud: server"
+sidebar_current: "docs-resource-server"
+description: |-
+  Creates and manages IonosCloud Server objects.
```
Comment (gabisavu-ionos) [inline docs/resources/cube_server.md:7]: IonosCloud C Server objects.

File: ionoscloud/data_source_cube_server.go [hunk 20]
```diff
@@ -0,0 +1,446 @@
...
+		ReadContext: dataSourceCubeServerRead,
+		Schema: map[string]*schema.Schema{
+			"template_uuid": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"datacenter_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.All(validation.StringIsNotWhiteSpace),
```
Comment (gabisavu-ionos) [inline ionoscloud/data_source_cube_server.go:24]: validateDiagFunc

File: ionoscloud/data_source_cube_server.go [hunk 21]
```diff
@@ -0,0 +1,446 @@
...
+			volumes = append(volumes, entry)
+		}
+
+		if err := d.Set("volumes", volumes); err != nil {
+			return err
+		}
+	}
+
+	var nics []interface{}
+	if server.Entities.Nics != nil && server.Entities.Nics.Items != nil && len(*server.Entities.Nics.Items) > 0 {
```
Comment (gabisavu-ionos) [inline ionoscloud/data_source_cube_server.go:305]: make a func which returns []interface{} and use it here and to the server.go as well

File: ionoscloud/resource_server_cube.go (hunk 14 above)
Comment (gabisavu-ionos) [reply ionoscloud/resource_server_cube.go:508]: I think i can use the func from server.go

File: ionoscloud/resource_server_cube.go [hunk 22]
```diff
@@ -0,0 +1,1336 @@
...
+		}
+
+		properties := ionoscloud.VolumeProperties{}
+
+		if v, ok := d.GetOk("volume.0.name"); ok {
+			vStr := v.(string)
+			properties.Name = &vStr
+		}
+
+		//if v, ok := d.GetOk("volume.0.size"); ok {
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_server_cube.go:963]: remove

File: ionoscloud/data_source_cube_server.go (hunk 9 above)
Comment (gabisavu-ionos) [reply ionoscloud/data_source_cube_server.go:34]: done

File: ionoscloud/data_source_cube_server.go [hunk 23]
```diff
@@ -0,0 +1,446 @@
...
+			}
+		}
+
+		if server.Properties.Name != nil {
+			if err := d.Set("name", *server.Properties.Name); err != nil {
+				return err
+			}
+		}
+
+		//if server.Properties.Cores != nil {
```
[the remark this answers is not in the document]
Comment (gabisavu-ionos) [reply ionoscloud/data_source_cube_server.go:196]: done

File: ionoscloud/resource_server_cube.go (hunk 10 above)
Comment (gabisavu-ionos) [reply ionoscloud/resource_server_cube.go:1336]: removed

File: ionoscloud/resource_server_cube.go (hunk 12 above)
Comment (gabisavu-ionos) [reply ionoscloud/resource_server_cube.go:1216]: removed

File: ionoscloud/resource_server_cube.go [hunk 24]
```diff
@@ -0,0 +1,1075 @@
...
+		}
+
+		if server.Properties.Name != nil {
+			if err := d.Set("name", *server.Properties.Name); err != nil {
+				diags := diag.FromErr(err)
+				return diags
+			}
+		}
+
+		//if server.Properties.Type != nil {
```
[the remark this answers is not in the document]
Comment (gabisavu-ionos) [reply ionoscloud/resource_server_cube.go:732]: removed

File: ionoscloud/resource_server_cube.go (hunk 13 above)
Comment (gabisavu-ionos) [reply ionoscloud/resource_server_cube.go:1157]: changed

File: ionoscloud/resource_server_cube.go (hunk 15 above)
Comment (gabisavu-ionos) [reply ionoscloud/resource_server_cube.go:37]: changed

File: docs/data-sources/cube_server.md [hunk 25]
```diff
@@ -0,0 +1,124 @@
+---
+subcategory: "Compute Engine"
+layout: "ionoscloud"
+page_title: "IonosCloud : ionoscloud_server"
```
[the remark this answers is not in the document]
Comment (gabisavu-ionos) [reply docs/data-sources/cube_server.md:4]: changed

File: docs/data-sources/cube_server.md (hunk 16 above)
Comment (gabisavu-ionos) [reply docs/data-sources/cube_server.md:7]: changed

File: docs/data-sources/cube_server.md (hunk 17 above)
Comment (gabisavu-ionos) [reply docs/data-sources/cube_server.md:12]: changed

File: docs/resources/cube_server.md (hunk 18 above)
Comment (gabisavu-ionos) [reply docs/resources/cube_server.md:36]: changed

File: docs/resources/cube_server.md (hunk 19 above)
Comment (gabisavu-ionos) [reply docs/resources/cube_server.md:7]: changed

File: docs/resources/cube_server.md [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+  datacenter_id     = ionoscloud_datacenter.example.id
+  public            = true
+  name              = "Lan Example"
+}
+
+resource "ionoscloud_server" "example" {
+  name              = "Server Example"
+  availability_zone = "ZONE_2"
+  image_name        = "ubuntu:latest"
+  type              = "CUBE"
```
[the remark this answers is not in the document]
Comment (gabisavu-ionos) [reply docs/resources/cube_server.md:40]: removed

File: ionoscloud/resource_server_cube.go (hunk 22 above)
Comment (gabisavu-ionos) [reply ionoscloud/resource_server_cube.go:963]: removed

File: ionoscloud/data_source_cube_server.go (hunk 20 above)
Comment (gabisavu-ionos) [reply ionoscloud/data_source_cube_server.go:24]: changed
